### PR TITLE
Unify cody configuration into single source of truth

### PR DESCRIPTION
**Pull request by eseliger**
This unified all cody configuration generation into a single source of truth and tests for the defaults settings for embeddings and completions we want for 5.1.

Ticks off a bunch of items in https://github.com/sourcegraph/sourcegraph/issues/52980.

## Test plan

Added a bunch of tests.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/site.go [hunk 1]
```diff
@@ -548,10 +549,11 @@ func (r *siteResolver) RequiresVerifiedEmailForCody(ctx context.Context) bool {
 	// App is typically forwarding Cody requests to dotcom.
```
Comment (chwarwick) [inline cmd/frontend/graphqlbackend/site.go:549]:
```suggestion

```

File: cmd/frontend/graphqlbackend/site.go [hunk 2]
```diff
@@ -548,10 +549,11 @@ func (r *siteResolver) RequiresVerifiedEmailForCody(ctx context.Context) bool {
...
 	if deploy.IsApp() {
-		// App users can specify their own keys
-		// if they use their own keys requests are not forwarded to dotcom
-		// which means a verified email is not needed
-		return conf.Get().Completions == nil
+		c := conf.GetCompletionsConfig(conf.Get().SiteConfig())
+		// App users can specify their own keys using one of the regular providers.
+		// If they use their own keys requests are not going through Cody Gateway
+		// which means a verified email is not needed.
+		return c != nil && c.Provider != conftypes.CompletionsProviderNameSourcegraph
```
Comment (chwarwick) [inline cmd/frontend/graphqlbackend/site.go:556]:
```suggestion
		return c == nil || c.Provider == conftypes.CompletionsProviderNameSourcegraph
```
It think this is backwards, App should require a verified email if it's using the default empty config or if the provider is `sourcegraph`  The dotcom user actor in the gateway checks if the user had verified emails and will return not enabled if there are no verified emails.

**Review by chwarwick — COMMENTED** [review]
This is such a nice clean up of all these settings thank you.

I think the `RequiresVerifiedEmailForCody` for app is off. It's a hard mind set in that, when running app that function is thinking about if dotcom requires a verified email even though the completions requests don't go though it anymore it's still a requirement of cody usage.

File: schema/site.schema.json [hunk 3]
```diff
@@ -2444,7 +2446,10 @@
         "enabled": {
           "description": "DEPRECATED. Use cody.enabled instead to turn Cody on/off.",
           "type": "boolean",
-          "default": false
+          "!go": {
+            "pointer": true
+          },
+          "default": true
```
Comment (bobheadxi) [inline schema/site.schema.json:2452]: Just to make sure I'm reading this right - the default is that both completions/embeddings are enabled whenever Cody is enabled? With the option to disable embeddings on top?

Comment (unknwon) [reply schema/site.schema.json:2452]: The "default" in the JSON schema is merely "default completion when user adds this field". Our Go code cares nothing about defaults here... 😂 (call it a feature or bug)

File: enterprise/cmd/cody-gateway/internal/actor/actor.go [hunk 4]
```diff
@@ -9,7 +9,7 @@ import (
 
 	"github.com/sourcegraph/sourcegraph/enterprise/cmd/cody-gateway/internal/limiter"
 	"github.com/sourcegraph/sourcegraph/enterprise/cmd/cody-gateway/internal/notify"
-	"github.com/sourcegraph/sourcegraph/enterprise/internal/codygateway"
+	"github.com/sourcegraph/sourcegraph/internal/codygateway"
```
Comment (unknwon) [inline enterprise/cmd/cody-gateway/internal/actor/actor.go:12]: :trollface:

File: enterprise/cmd/cody-gateway/internal/actor/productsubscription/productsubscription.go [hunk 5]
```diff
@@ -16,9 +16,10 @@ import (
 
 	"github.com/sourcegraph/sourcegraph/enterprise/cmd/cody-gateway/internal/actor"
 	"github.com/sourcegraph/sourcegraph/enterprise/cmd/cody-gateway/internal/dotcom"
-	"github.com/sourcegraph/sourcegraph/enterprise/internal/codygateway"
-	"github.com/sourcegraph/sourcegraph/enterprise/internal/licensing"
+	elicensing "github.com/sourcegraph/sourcegraph/enterprise/internal/licensing"
 	"github.com/sourcegraph/sourcegraph/enterprise/internal/productsubscription"
+	"github.com/sourcegraph/sourcegraph/internal/codygateway"
+	"github.com/sourcegraph/sourcegraph/internal/licensing"
```
Comment (unknwon) [inline enterprise/cmd/cody-gateway/internal/actor/productsubscription/productsubscription.go:22]: New import but does nothing?

File: enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go [hunk 6]
```diff
@@ -29,7 +30,7 @@ func NewAnthropicHandler(
 		eventLogger,
 		rs,
 		rateLimitNotifier,
-		anthropic.ProviderName,
+		string(conftypes.CompletionsProviderNameAnthropic),
```
Comment (unknwon) [inline enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go:33]: Could we make the `makeUpstreamHandler` accepts `conftypes.CompletionsProviderName` instead? feel more strongly typed.

File: enterprise/cmd/frontend/internal/embeddings/resolvers/resolvers_test.go [hunk 7]
```diff
@@ -21,6 +21,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/featureflag"
 	"github.com/sourcegraph/sourcegraph/internal/gitserver"
 	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/pointers"
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/embeddings/resolvers/resolvers_test.go:24]: TIL we have this. cc @kopancek IIRC you wanted a lib for making pointers? Ignore me if my memory messed up 😅

File: enterprise/internal/embeddings/background/repo/store_test.go [hunk 8]
```diff
@@ -365,3 +376,7 @@ func createGlobalPolicy(ctx context.Context, store RepoEmbeddingJobsStore) error
 	)
 	return store.Exec(ctx, q)
 }
+
+func pointify[T any](v T) *T {
```
Comment (unknwon) [inline enterprise/internal/embeddings/background/repo/store_test.go:380]: Don't we have `pointers.Ptr`?

File: enterprise/internal/licensing/check.go [hunk 9]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/conf"
 	"github.com/sourcegraph/sourcegraph/internal/goroutine"
 	"github.com/sourcegraph/sourcegraph/internal/httpcli"
+	olicensing "github.com/sourcegraph/sourcegraph/internal/licensing"
```
Comment (unknwon) [inline enterprise/internal/licensing/check.go:19]: Import alias not actually needed?

**Review by unknwon — APPROVED** [review]

File: enterprise/cmd/frontend/internal/embeddings/resolvers/resolvers_test.go (hunk 7 above)
Comment (unknwon) [reply enterprise/cmd/frontend/internal/embeddings/resolvers/resolvers_test.go:24]: Oh... [you made it](https://github.com/sourcegraph/sourcegraph/pull/53441).

File: schema/site.schema.json (hunk 3 above)
Comment (bobheadxi) [reply schema/site.schema.json:2452]: I thought JSON schema "default completion" is supposed to be "examples"? 😬 How else do we document the implied default to the user? 😅

Comment (unknwon) [reply schema/site.schema.json:2452]: Depending on the field type, could be the default selected value, or just get auto-completes.

CleanShot 2023-06-14 at 19 39 00@2x

https://github.com/sourcegraph/sourcegraph/assets/2946214/48772be3-4e5b-4c41-be29-1df5d88940fb

File: enterprise/internal/embeddings/background/repo/store_test.go (hunk 8 above)
Comment (eseliger) [reply enterprise/internal/embeddings/background/repo/store_test.go:380]: heh, that was introduced during some of the merges I did from upstream.. 😆

File: schema/site.schema.json (hunk 3 above)
Comment (eseliger) [reply schema/site.schema.json:2452]: > The "default" in the JSON schema is merely "default completion when user adds this field". Our Go code cares nothing about defaults here... so in Go, the default is the zero value of bool, which is still false (haven't checked other files in this PR) 😂 Call it a feature or bug.


True, I tried to align what we actually use as the default in code and what we advertise as the default in the json schema :) 

And to Roberts question: yes, cody.enabled will enable both, but you can optionally disable either.

File: enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go (hunk 6 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go:33]: I was very hesitant to import from conf, even if it's just a subpackage that (at the moment) isn't doing anything harmful.. 😬

Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go:33]: It's `conftypes` not `conf` though, like only has bunch of shared types.

Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go:33]: yes it currently doesn't but you know.. who guards that 😆 I mean I can do this (to then convert back to string when we pass it to the logger) but it felt a little bit scary to me 😬

File: enterprise/cmd/cody-gateway/internal/actor/productsubscription/productsubscription.go (hunk 5 above)
Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/actor/productsubscription/productsubscription.go:22]: it does, it is used wherever enterprise/licensing was used before :)

File: cmd/frontend/graphqlbackend/site.go (hunk 2 above)
Comment (eseliger) [reply cmd/frontend/graphqlbackend/site.go:556]: thanks, fixed!

File: enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go (hunk 6 above)
Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/httpapi/completions/anthropic.go:33]: Alright it's a nit after all...!

**Pull request merged**